osu.Game.Rulesets.Osu/Difficulty/Evaluators/AimEvaluator.cs

```diff
@@ -33,8 +39,46 @@ public static double EvaluateDifficultyOf(DifficultyHitObject current, bool with
             var osuLastObj = (OsuDifficultyHitObject)current.Previous(0);
             var osuLastLastObj = (OsuDifficultyHitObject)current.Previous(1);
 
+            // Doubletap detection stuff
+
+            // Calculate how much things are overlapping. Low overlapping circles are very hard in being doubletappable
+            double overlapnessCurr = Math.Clamp(osuCurrObj.LazyJumpDistance / OsuDifficultyHitObject.NORMALISED_RADIUS / 2, 0, 1);
+            overlapnessCurr = CosSigmoid(overlapnessCurr);
+            overlapnessCurr *= overlapnessCurr;
+
+            double antiOverlapnessLast = 1 - Math.Clamp(osuLastObj.LazyJumpDistance / OsuDifficultyHitObject.NORMALISED_RADIUS / 2, 0, 1);
+            antiOverlapnessLast = CosSigmoid(antiOverlapnessLast);
+            antiOverlapnessLast *= antiOverlapnessLast;
+
+            // Doubletap hitwindow lands in range [0, 300 hitwindow], where 0 is impossible to doubletap 
+            double doubletapHitWindow = (osuLastObj.HitWindowGreat + osuLastLastObj.HitWindowGreat) / 2 - osuLastObj.StrainTime;
+            
+            // Extra time to aim when doubletapping
+            double doubletapTime = 0;
+
+            if (doubletapHitWindow > 0)
+            {
+                // Always equal or bigger than doubletapTime from `else`
+                doubletapTime = osuLastObj.StrainTime * overlapnessCurr * antiOverlapnessLast / 2;
+            }
+            else 
+            {
+                doubletapTime = osuLastObj.HitWindowGreat * overlapnessCurr * antiOverlapnessLast / 2;
+
+                // if the second note is delayed - decrease the penalty down to 0 on doubled time
```

Comment:
I'm still not sure if this is something we wanna do, but I'll wait for the sheet before deciding

Reply:
this is questionable because it assumes that player will click in the first 1 ms of the 300 window, so it can be adjusted to be more fair
but the core is very logical to me